src/obfs/auth_chain.c

```diff
@@ -261,50 +657,148 @@ void auth_chain_a_dispose(obfs *self)
+                    key_change_interval = (uint64_t) n;
+                }
+            }
+        }
+    }
+
+    uint8_t *key_change_datetime_key_bytes = (uint8_t *) malloc(sizeof(uint8_t) * 8);
+    uint64_t key_change_datetime_key = (uint64_t) (time(NULL)) / key_change_interval;
+    for (int i = 7; i >= 0; --i) {
+        key_change_datetime_key_bytes[7 - i] = (uint8_t) ((key_change_datetime_key >> (8 * i)) & 0xFF);
+    }
+
+    auth_chain_f_init_data_size(self, &self->server, key_change_datetime_key_bytes);
+
+    free(key_change_datetime_key_bytes);
+    key_change_datetime_key_bytes = NULL;
+}
+
+unsigned int auth_chain_udp_get_rand_len(shift128plus_ctx *random, uint8_t *last_hash) {
     shift128plus_init_from_bin(random, last_hash, 16);
     return shift128plus_next(random) % 127;
 }
 
-unsigned int get_rand_start_pos(int rand_len, shift128plus_ctx *random)
-{
+unsigned int auth_chain_get_rand_start_pos(int rand_len, shift128plus_ctx *random) {
     if (rand_len > 0)
         return shift128plus_next(random) % 8589934609 % rand_len;
     return 0;
 }
 
-unsigned int get_client_rand_len(auth_chain_local_data *local, int datalength, int overhead)
-{
-    return local->rnd(datalength, &local->random_client, local->last_client_hash, local->data_size_list, local->data_size_list_len,
-                      local->data_size_list2, local->data_size_list2_len, overhead);
+unsigned int auth_chain_get_client_rand_len(auth_chain_local_data *local, server_info_t *server, int datalength) {
+    return local->get_tcp_rand_len(local, server, datalength, &local->random_client, local->last_client_hash);
 }
 
-unsigned int get_server_rand_len(auth_chain_local_data *local, int datalength, int overhead)
-{
-    return local->rnd(datalength, &local->random_server, local->last_server_hash, local->data_size_list, local->data_size_list_len,
-                      local->data_size_list2, local->data_size_list2_len, overhead);
+unsigned int auth_chain_get_server_rand_len(auth_chain_local_data *local, server_info_t *server, int datalength) {
+    return local->get_tcp_rand_len(local, server, datalength, &local->random_server, local->last_server_hash);
 }
 
-int auth_chain_a_pack_data(char *data, int datalength, char *outdata, auth_chain_local_data *local, server_info_t *server)
-{
-    unsigned int rand_len = get_client_rand_len(local, datalength, server->overhead);
+int auth_chain_a_pack_data(char *data, int datalength, char *outdata, auth_chain_local_data *local,
+                           server_info_t *server) {
+    unsigned int rand_len = auth_chain_get_client_rand_len(local, server, datalength);
     int out_size = (int) rand_len + datalength + 2;
     outdata[0] = (char) ((uint8_t) datalength ^ local->last_client_hash[14]);
     outdata[1] = (char) ((uint8_t)(datalength >> 8) ^ local->last_client_hash[15]);
 
     {
-        VLA(uint8_t, rand_len, rnd_data);
+        uint8_t * rnd_data = (uint8_t *)malloc(rand_len);
```

Comment:
The VLA macro may get benefit from GCC or clang. It's not a good idea replace it with malloc.